conans/client/downloaders/cached_file_downloader.py

```diff
@@ -1,75 +1,68 @@
-                    self._file_downloader.download(url=url, file_path=cached_path, md5=md5,
-                                                   sha1=sha1, sha256=sha256, **kwargs)
-                else:
-                    # specific check for corrupted cached files, will raise, but do nothing more
-                    # user can report it or "rm -rf cache_folder/path/to/file"
-                    try:
-                        check_checksum(cached_path, md5, sha1, sha256)
-                    except ConanException as e:
-                        raise ConanException("%s\nCached downloaded file corrupted: %s"
-                                             % (str(e), cached_path))
-
-                if file_path is not None:
-                    file_path = os.path.abspath(file_path)
-                    mkdir(os.path.dirname(file_path))
-                    shutil.copy2(cached_path, file_path)
-                else:
-                    with open(cached_path, 'rb') as handle:
-                        tmp = handle.read()
-                    return tmp
+                yield
             finally:
                 thread_lock.release()
 
-    def _get_hash(self, url, checksum=None):
-        """ For Api V2, the cached downloads always have recipe and package REVISIONS in the URL,
-        making them immutable, and perfect for cached downloads of artifacts. For V2 checksum
-        will always be None.
-        For ApiV1, the checksum is obtained from the server via "get_snapshot()" methods, but
-        the URL in the apiV1 contains the signature=xxx for signed urls, but that can change,
-        so better strip it from the URL before the hash
+    def download(self, url, file_path=None, md5=None, sha1=None, sha256=None, **kwargs):
+        """ compatible interface of FileDownloader + checksum
         """
-        scheme, netloc, path, _, _ = urlsplit(url)
-        # append empty query and fragment before unsplit
-        if not self._user_download:  # removes ?signature=xxx
-            url = urlunsplit((scheme, netloc, path, "", ""))
-        if checksum is not None:
-            url += checksum
-        h = sha256_sum(url.encode())
-        return h
+        checksum = sha256 or sha1 or md5
+        # If it is a user download, it must contain a checksum
+        assert (not self._user_download) or (self._user_download and checksum)
+        h = hash_url(url, checksum, self._user_download)
+
+        with self._lock(h):
+            cached_path = os.path.join(self._cache_folder, h)
+            if not os.path.exists(cached_path):
+                self._file_downloader.download(url=url, file_path=cached_path, md5=md5,
+                                               sha1=sha1, sha256=sha256, **kwargs)
+                with self._lock(self._cache_mapping):
+                    mapping_filename = os.path.join(self._cache_folder, self._cache_mapping)
+                    with open(mapping_filename, "a") as f:
```

Comment:
Maybe it would be better to explicitly store the "user-download" in this data file? Are you going to filter later based on URLs only?

Reply:
Yes, the filter uses the URLs from the remotes (implemented here: https://github.com/conan-io/conan-jfrog-cli-plugin/blob/87eacc796918784e013a5aefbf31a2885228cc25/commands/backup_sources.go#L140).

Anyway, there was a fortunate coincidence, as usage of the downloaders from our rest clients (calls to Conan remotes) and from `tools.get/download` is totally isolated, this _ArtifactoryCache_ middleware is only added for the `tools.get/download` calls, all of them should probably _user_downloads_.